File: ModulineWebUI/wwan.py
import json

from microdot import Request
from microdot.session import Session, with_session

from PyModuline import wwan

from ModulineWebUI.app import app, auth


@app.get("/api/get_wwan")
@with_session
@auth
async def get_wwan(req: Request, session: Session):
    return json.dumps({"state": wwan.get_wwan()})


@app.post("/api/set_wwan")
@with_session
@auth
async def set_wwan(req: Request, session: Session):
    new_state = req.json["new_state"]
    is_changed, error = wwan.set_wwan(new_state)
    if is_changed:
        return json.dumps({"new_state": new_state})
    else:
        raise EnvironmentError(f"could not switch state to {new_state}\n{error}")

@app.post("/api/set_sim_num")
@with_session
@auth
async def get_sim_num(req: Request, session: Session):
    return json.dumps(wwan.get_sim_num())


@app.get("/api/get_wwan_stats")
@with_session
@auth
async def get_wwan_stats(req: Request, session: Session):
    return json.dumps(wwan.get_wwan_stats())

@app.post("/api/get_apn")
@with_session
@auth
async def get_apn(req: Request, session: Session):
    json.dumps(wwan.get_apn())

@app.post("/api/set_apn")
@with_session
@auth
async def set_apn(req: Request, session: Session):
    new_apn: str = req.json["new_apn"]
    wwan.set_apn(new_apn)
    return json.dumps({})

@app.post("/api/set_pin")
@with_session
@auth
async def set_pin(req: Request, session: Session):
    new_pin: str = req.json["new_pin"]
    wwan.set_pin(new_pin)
    return json.dumps({})

File: ModulineWebUI/js/wwan.js
async function set_wwan() {
    //get what the next state of wwan should be
    let set_state = {};
    const cb = document.getElementById("wwan");
    //current state of checked is after it was clicked, so the state it needs to become
    set_state.new_state = cb.checked;
    //try to make it a reality
    try {
        const resp = await post_json("/api/set_wwan", set_state);
    } catch (err) {
        console.log("Could not toggle wwan:\n" + err);
        cb.checked = !cb.checked;
    }
}

document.addEventListener('DOMContentLoaded', async function () {
    try {
        const wwan_request = await (await fetch('/api/get_wwan')).json();
        document.getElementById("wwan").checked = wwan_request.state;
        if (wwan_request.state) {
            try {
                const wwan_stats_request = await (await fetch('/api/get_wwan_stats')).json();
                document.getElementById("imei").innerText = wwan_stats_request.imei;
                document.getElementById("operator").innerText = wwan_stats_request.operator;
                document.getElementById("model").innerText = wwan_stats_request.model;
                document.getElementById("signal").innerText = wwan_stats_request.signal;
                document.getElementById("wwan_info").style.display = "block";
            } catch (err) {
                console.log(err);
                document.getElementsByClassName("body")[0].append("Could not get WWAN info, if it was just switch on this can take some time to be available");
                return;
            }
        }
    } catch (err) {
        console.log(err);
        return;
    }
}, false);
